addon/-private/system/store.js

```diff
@@ -1926,50 +1912,56 @@ Store = Service.extend({
+    assert(`'${internalModel.modelName}' was saved to the server, but the response does not have an id and your record does not either.`, !(id === null && oldId === null));
 
     // ID absolutely can't be different than oldID if oldID is not null
-    assert(`'${internalModel.type.modelName}:${oldId}' was saved to the server, but the response returned the new id '${id}'. The store cannot assign a new id to a record that already has an id.`, !(oldId !== null && id !== oldId));
+    assert(`'${internalModel.modelName}:${oldId}' was saved to the server, but the response returned the new id '${id}'. The store cannot assign a new id to a record that already has an id.`, !(oldId !== null && id !== oldId));
 
     // ID can be null if oldID is not null (altered ID in response for a record)
     // however, this is more than likely a developer error.
     if (oldId !== null && id === null) {
-      warn(`Your ${internalModel.type.modelName} record was saved to the server, but the response does not have an id.`, !(oldId !== null && id === null));
+      warn(`Your ${internalModel.modelName} record was saved to the server, but the response does not have an id.`, !(oldId !== null && id === null));
       return;
     }
 
-    this.typeMapFor(internalModel.type).idToRecord[id] = internalModel;
+    this._recordMapFor(internalModel.modelName).idToRecord[id] = internalModel;
 
     internalModel.setId(id);
   },
 
   /**
-    Returns a map of IDs to client IDs for a given type.
+   Returns the normalized (dasherized) modelName. This method should be used whenever
+   receiving a modelName in a public method.
+
 
-    @method typeMapFor
+   @method _classKeyFor
+   @param {String} modelName
+   @returns {String}
+   @private
+   */
+  _classKeyFor(modelName) {
+    return normalizeModelName(modelName);
+  },
+
+  /**
+    Returns a map of IDs to client IDs for a given modelName.
+
+    @method _recordMapFor
     @private
-    @param {DS.Model} modelClass
-    @return {Object} typeMap
+    @param {String} modelName
+    @return {Object} recordMap
   */
-  typeMapFor(modelClass) {
-    heimdall.increment(typeMapFor);
-    let typeMaps = get(this, 'typeMaps');
-    let guid = guidFor(modelClass);
-    let typeMap = typeMaps[guid];
-
-    if (typeMap) { return typeMap; }
-
-    heimdall.increment(typeMapFor_allocate);
-    typeMap = {
-      idToRecord: new EmptyObject(),
-      records: [],
-      metadata: new EmptyObject(),
-      type: modelClass
-    };
+  _recordMapFor(modelName) {
+    heimdall.increment(_recordMapFor);
+    let recordMaps = this._recordMaps;
+    let recordMap = recordMaps[modelName];
 
-    typeMaps[guid] = typeMap;
+    if (recordMap) { return recordMap; }
 
-    return typeMap;
+    recordMap = new ClassMeta(this, modelName);
+    recordMaps[modelName] = recordMap;
```

Comment:
I originally was going to pass in `owner`, but wanted to piggy-pack on the `modelFor` behavior (for now). I think longer term `modelFor` would move into this object itself.